### ISSUE TITLE
troubles with autodiscover with the new outlook 2016 assistant

### ISSUE THREAD

**Issue by mritzmann** [mailcow/mailcow-dockerized#615]
Outlook 2016 has problems with the autodiscover function of Mailcow. The old setup wizard works, the new one does not. Anybody got the same problem?

---

new wizard:

![new_assistant](https://user-images.githubusercontent.com/4533780/30533220-881337e0-9c58-11e7-8ae9-75c7f8a375aa.png)

old wizard:

![old_assistant](https://user-images.githubusercontent.com/4533780/30533229-911aa01c-9c58-11e7-9e5e-06388f223f98.png)

The configuration looks valid to me:

![test](https://user-images.githubusercontent.com/4533780/30533245-a550091e-9c58-11e7-9c86-e588415fbf13.png)

**Comment by mkuron** [comment]
Please use the old one for now. Outlook 2016 for Mac has always shown the same issue (it uses a new JSON-based autodiscover mechanism that is not documented by Microsoft and not supported by us), so I suspect they ported that to the Windows version now.

**Comment by andryyy** [comment]
Sounds fun.

Would love to wireshark that communication.

**Comment by mritzmann** [comment]
If anyone is interested in debugging: I can donate an Office365 account with Outlook 2016.

**Comment by ChessSpider** [comment]
quick way to open old style:
in RUN
```C:\Program Files (x86)\Microsoft Office\root\Office16\OLCFG.EXE```
then button ```email accounts``` => ```manual setup or additional server types``` => **Exchange ActiveSync**

I had to look for this, as autoconfigure on Outlook 2016 now configures as SMTP/IMAP :(

**Comment by andryyy** [comment]
@mkuron Any idea why it configures as IMAP?

**Comment by mritzmann** [comment]
> I had to look for this, as autoconfigure on Outlook 2016 now configures as SMTP/IMAP

Tested it yesterday with the old setup-assistant. Account was set up as ActiveSync. Only the new setup-assistant doesn't work.

**Comment by andryyy** [comment]
Strange. But it auto-configured it as IMAP?

**Comment by mritzmann** [comment]
automatically as activesync

**Comment by mkuron** [comment]
I'll test and fix it later this week. The IMAP thing is probably a user agent issue again.

**Comment by andryyy** [comment]
@mritzmann I meant the new config guide. :) Thanks mkuron! :)

**Comment by mkuron** [comment]
I haven't had any success with Outlook 2016's autoconfiguration. The OLCFG.EXE trick suggested above doesn't work for me as it always asks for my admin password and then launches as Administrator, not the user I'm running Outlook as. Even worse, when I click the "New" button, it gives me the new wizard. I am simply unable to get to the old wizard.

The new wizard leads to a single request to https://autodiscover.example.com/autodiscover/autodiscover.json?Email=user%40example.com&Protocol=ActiveSync&RedirectCount=1 with user agent _OutlookMobileCloudService-Autodetect/1.0.0_ -- and that request is sent by some Microsoft IP address, not the Outlook client itself. I've implemented the JSON protocol and will post a PR soon. Unfortunately though, I'm not actually able to get Outlook to properly configure the account quite yet.

If I manually create an Exchange account with the new wizard, I get the request to autodiscover.xml as in older Outlook versions, with a user agent of _Microsoft Office/16.0 (Windows NT 10.0; MAPI 16.0.8431; Pro)_. The autodiscover.php script handles that correctly and sends ActiveSync configuration if that hasn't been disabled. Unfortunately though, Outlook fails to properly deal with the configuration it receives.

The best thing I currently have is outlook.com's reply to autodiscovery requests, e.g. at http://autodiscover.outlook.com//autodiscover/autodiscover.json?Email=test%40outlook.com&Protocol=ActiveSync&RedirectCount=1 . If you change the protocol to something else, the error message will contain a list of all supported protocols.

For now, the executive summary is: autodiscover is broken in Outlook 2016 with anything but Office 365 (even Microsoft Exchange Server appears to have trouble in many setups -- I tried two and only one worked). I can't fix it until Microsoft publishes documentation or someone reverse engineers the full protocol.

**Comment by mritzmann** [comment]
Hi @mkuron You dealt with it a lot in September. Thanks for that! Is there already an update? Has Microsoft published any specifications yet?

**Comment by RickD123** [comment]
Hi,

does anyone have the specifications or at least a sample of the autodiscover.json file format used by Outlook? Anyone used Wireshark on this yet?

Thanks!

Kind regards
RickD

**Comment by mkuron** [comment]
There is no public spec. https://github.com/mailcow/mailcow-dockerized/blob/master/data/web/autodiscover-json.php is based on samples we got by querying Hotmal‘s autodiscover.json and different Exchange Server versions. It is incomplete though — it seems like either Outlook performs some extra magic during discovery, or recent Outlook/Exchange versions don’t really support autodiscovering anything but EAS and MAPI.

**Comment by amaccuish** [comment]
Hi. I've just been looking into this and I found a site (in german) that describes the new json protocol quite well. Here you go: https://www.msxfaq.de/exchange/autodiscover/autodiscover_v2.htm

**Comment by mkuron** [comment]
Thanks @amaccuish, but unfortunately it doesn't contain any information we didn't already figure out ourselves.

**Comment by amaccuish** [comment]
@mkuron ahh no worries. I used it for a small script that seems to work fine with office 2016, thought it could be useful :)

**Comment by graczu0x0x0x0x0** [comment]
Any success on compatibility for new Outlook?

**Comment by jol64** [comment]
I am also affected... autodiscovery works with Outlook 2013 but fails with Outlook 365.

**Comment by andryyy** [comment]
Outlook from Office 365 is shit. I doubt they even have EAS anymore. See this issue and others mentioning 365.

**Comment by Shadowghost** [comment]
No idea about Office 365, but with the normal Office 2016 setup wizard, Exchange seems to work fine, mails, contacts and calendar is synced. Tried it today with latest master and properly configured DNS.

**Comment by jol64** [comment]
Office 365 is a trend - I got it with the Aldi convertible. I dislike the idea of switching to Thunderbird and will also have a hard time to explain my users that Office 365 cannot be used with Mailcow.

**Comment by andryyy** [comment]
I don't think so.

We cannot fix Outlok 365 anyway. So if you want to use EAS with Outlook, don't buy Office 365. Just use Outlook CalDav Synchronizer.

**Comment by jol64** [comment]
I googled once more today and found the following interesting links:
https://support.microsoft.com/en-us/help/3211279/outlook-2016-implementation-of-autodiscover
https://www.msxfaq.de/exchange/autodiscover/autodiscover_v2.htm
these might help to track down the issue.
plus
https://forum.netcup.de/administration-eines-server-vserver/vserver-server-kvm-server/p94909-office365-business-essentials-durch-mailcow-auf-rs1000-g8-ersetzen-eure-tipps/#post94909
gives a description how to manually configure a mailbox and that worked for me. Short summary; open control panel, search for mail, autoconfig will probably fail, then pick configure manually.

**Comment by entr0p1** [comment]
Can confirm on the latest release of O365 (I think matches up with Office 2019) that Outlook won't allow the account to be created as an Exchange account, however going through the "Control Panel -> Mail -> Email Accounts..." and adding it from there works fine. Hope this helps.

**Comment by Individuum92** [comment]
I tried Office 365 and Outlook 2016 and both dont work for me.
Im new with Mailcow. Is there any fix for this issue?

**Comment by andryyy** [comment]
The retail Outlooks work. They have a "normal" autodiscovery and do not lack EAS. 

There is no fix for something we cannot change. Don't use stupid Outlook. And don't use EAS on a desktop at all. :/

**Comment by rsecor** [comment]
Any news on this? This has been causing my clients nightmares!

**Comment by andryyy** [comment]
No. I don't own Outlook and don't plan to buy that crap anytime soon. 

Contributions are welcome though.

**Comment by mathisloge** [comment]
Just as a small sidenote: 

just tried it with the latest office365 and with the setup over `C:\Program Files\Microsoft Office\root\Office16\OLCFG.EXE` it setups the active sync without problems

**Comment by Adorfer** [comment]
> Any news on this?

works on office2016 and office365.
(setting up new clients regularly. can't see any issue with mailcow.) 

P. S. in case you have trouble setting up your dns according to the docs it would be a support question out of scope here.

**Comment by RickD123** [comment]
This bug is still a problem and I could not configure an IMAP account in Office 365. It was only after doing a "reg add HKEY_CURRENT_USER\SOFTWARE\Policies\Microsoft\Office\16.0\Outlook\setup /T REG_DWORD /F /V DisableOffice365SimplifiedAccountCreation /D 1" that I was able to manually create the account.

**Comment by andryyy** [comment]
Why were you not able to create an IMAP account? That's never an issue.

If you use a stupid Office 365 version, which will only allow you to add an Office 365 account, I don't feel the need to change anything in mailcow. You should switch to a proper IMAP client then. :)

**Comment by Pihaar** [comment]
There is a website from microsft which emutaltes the access of outlook to various Mail Server Services:
https://testconnectivity.microsoft.com/tests/exchange

E.g. for ActiveSync I got this result:
![image](https://user-images.githubusercontent.com/2302159/81666549-7c287500-9442-11ea-8e7f-9b7cb15bce5d.png)

Here the error message in plain text:
```
Testing HTTP Authentication Methods for URL https://mail.example.com/Microsoft-Server-ActiveSync.
The HTTP authentication test failed.
A 401 error was received from the server, but no authentication methods are supported. HTTP Response Headers: Connection: keep-alive X-Frame-Options: SAMEORIGIN Content-Length: 0 Content-Type: text/plain; charset=utf-8 Date: Tue, 12 May 2020 07:52:42 GMT Server: nginx WWW-Authenticate: basic realm="SOGo"
Elapsed Time: 483 ms.
```

The shown SSL warnings are not relevant in my oppinion.

You can try this out, but I'd recommend not to use your productive mailbox as you need to provide your credentials.

**Comment by rsecor** [comment]
@Pihaar Unfortunately that tool does not mimic reality when it comes to Microsoft Outlook 365 - Desktop Version.

**Comment by J-Rey** [comment]
Microsoft 365 (O365) [deprecated EAS in favor of REST a few years ago](https://techcommunity.microsoft.com/t5/outlook/deprecating-eas-support-for-outlook-for-ios-amp-android/m-p/63799) but that's irrelevant since clients should expect to connect to Mailcow the same as Exchange Server which still enables EAS by default.

The issue is that Mailcow is using the oldest [method of detecting via Autodiscover](https://docs.microsoft.com/en-us/exchange/client-developer/web-service-reference/autodiscover-web-service-reference-for-exchange) that was for Exchange 2007, **POX**.

If we _at least_ use **SOAP** for Autodiscover service (introduced [with Exchange 2010](https://docs.microsoft.com/en-us/exchange/client-developer/exchange-web-services/autodiscover-for-exchange#options-for-using-autodiscover)) then maybe we won't have to use [the workaround](https://github.com/mailcow/mailcow-dockerized-docs/pull/62) & could support Outlook for macOS by the time they [support on-premises Exchange later this year](https://support.microsoft.com/en-us/office/the-new-outlook-for-mac-6283be54-e74d-434e-babb-b70cefc77439?ui=en-us&rs=en-us&ad=us).

To be thorough, we could also consider supporting the [primary API](https://docs.microsoft.com/en-us/exchange/client-developer/web-service-reference/web-services-reference-for-exchange) for Exchange, **EWS**. It does have [Autodiscover](https://docs.microsoft.com/en-us/exchange/client-developer/exchange-web-services/autodiscover-for-exchange#options-for-using-autodiscover) but its schema changes for each version of Exchange Server & they promote a managed API so it seems not to be worth the extra effort especially considering how slow this issue has been progressing.

**Comment by mkuron** [comment]
Finally, Microsoft provides some documentation on this... Unfortunately, judging by https://docs.microsoft.com/en-us/exchange/client-developer/web-service-reference/setting-soap, it seems like the SOAP autodiscovery only works for EWS. As long as SOGo doesn't support EWS, it won't be useful for Mailcow.
